### DISCUSSION TITLE
Per git op client?

### DISCUSSION THREAD

**Discussion by Madoxen** [go-git/go-git#1126] [Q&A] [answered]
I know that one can override a default http client for git operations by doing `InstallProtocol` as in https://github.com/go-git/go-git/blob/master/_examples/custom_http/main.go. However when I've inspected `InstallProtocol` source it occurred to me that this function modifies a package-global variable that holds clients for various schemes. 

Normally this is not a problem but what if I want to have two differently configured `http.Client` for the same scheme (like `https`) and I want to run these independently of each other (e.g. in different goroutines)? 

Is there some way to declare client override per one git operation (like `List` or `Clone`)? Switching between clients with `InstallProtocol` would require global locks, which makes concurrent use a bit useless.

**Comment by pjbgf** [comment, marked as answer]
@Madoxen What exactly do you need the different clients to do? One way around this could be to simply register arbitrary schemes: e.g. `unsafe://` bound to a custom http client that has TLS verification disabled for example.

**Comment by Madoxen** [comment]
I probably could do that, but this involves rewriting the scheme later in the client, most probably in round tripper. This seems fragile to me. I would really like to pass a custom client as a parameter to remote or a function that needs it, instead of relying on `Protocols` global :/

Anyways thanks for help, if there is any other way to do this please let me know.
